ckanext/unaids/actions.py

```diff
@@ -264,3 +264,61 @@ def package_create(next_action, context, data_dict):
 
 def time_ago_from_timestamp(context, data_dict):
     return t.h.time_ago_from_timestamp(data_dict.get('timestamp'))
+
+
+def dataset_lock(context, data_dict):
+    _check_access('dataset_lock', context, data_dict)
+    dataset_id = t.get_or_bust(data_dict, "id")
+    if not t.h.dataset_lockable(dataset_id):
+        raise t.ValidationError(t._("Datasets of this type cannot be locked."))
+    locked_name = t.config.get("ckanext.unaids.locked_release_name", "Locked")
+    context['bypass_read_only'] = True
+    t.get_action("package_patch")(
+        context,
+        {"id": dataset_id, "locked": True}
+    )
+    try:
+        t.get_action("dataset_version_create")(context, {
+            "dataset_id": dataset_id,
+            "name": locked_name,
+            "notes": t._("This dataset has been made read-only by a system administrator")
+        })
+    except t.ValidationError:
```

Comment:
The check if dataset version with name `locked_name` can be created be better suited for the helper `h.dataset_lockable(dataset_id)`
In case this rollback happens you'll end up with polluted activity stream with:
1. "locked" = true
2. "locked" = false

Reply:
Yes I see the point you are making.  I think it doesn't fit in `dataset_lockable` perhaps because it really needs a subtly different name to the helper "dataset_type_lockable" or some such name. But we could develop some separate function that checks the releases list prior to patching the dataset.